AngularApp/src/app/diagnostic-data/services/telemetry/telemetry.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable, OnInit, Inject } from '@angular/core';
+import { TelemetryEventNames, ITelemetryProvider } from './telemetry.common';
+import { DIAGNOSTIC_DATA_CONFIG, DiagnosticDataConfig } from '../../config/diagnostic-data-config';
+import { AppInsightsTelemetryService } from './appinsights-telemetry.service';
+import { KustoTelemetryService } from './kusto-telemetry.service';
+import { BehaviorSubject } from 'rxjs';
+
+@Injectable()
+export class TelemetryService {
+    private telemetryProviders: ITelemetryProvider[] = [];
+    private commonDetectorEventProperties: { [name: string]: string };
+    eventPropertiesSubject: BehaviorSubject<any> = new BehaviorSubject<any>(null);
+    eventPropertiesLocalCopy: { [name: string]: string };
+
+    constructor(private _appInsightsService: AppInsightsTelemetryService, private _kustoService: KustoTelemetryService, @Inject(DIAGNOSTIC_DATA_CONFIG) private config: DiagnosticDataConfig) {
+        if (config.useKustoForTelemetry) {
+            this.telemetryProviders.push(this._kustoService);
+        }
+        if (config.useAppInsightsForTelemetry) {
+            this.telemetryProviders.push(this._appInsightsService);
+        }
+
+        this.eventPropertiesSubject.subscribe((data: any) => {
+            this.eventPropertiesLocalCopy = data;
+        });
+    }
+
+    public setDetectEventProperties(eventProperties: { [name: string]: string }) {
```

Comment:
Instead of this function which is Detector specific in this generic service, can we use the about eventPropertiesSubject instead. 
Right now, you are overwriting the eventPropertiesLocalCopy with whatever gets pushed to the Behavior Subject. Instead of that, if you  update the keys that are present in data to this eventPropertiesLocalCopy  (Like you are doing in LogEvent Method), then the  detector columns (start time, end time  and detector id) can also be pushed to this Behavior Subject and you can get rid of this method..

Also, it will be a consistent way to communicate any common properties that we want to write along with every event.. Thoughts?

Reply:
Yes I am confused about the difference between eventPropertiesSubject/eventPropertiesLocalCopy  and commonDetectorEventProperties. Doesn't look like you are actually using the subject anywhere yet? Are they both just trying to solve the problem of common columns that you want to log in every logEvent?

Reply:
Yes for the consistency, I totally agree we should have all the data flowing in the same manner by pushing into eventPropertiesSubject. What Shekhar is suggesting is that we can updating the eventPropertiesLocalCopy key value pairs each time instead of overwriting, I will fix that for the current scenario and consistency. Do you think  there might be other scenarios besides detectors where this telemetry service can be applied? We probably don't want the detector common columns for that, right?